pandas/core/frame.py

```diff
@@ -893,8 +893,8 @@ def __dataframe__(
         Parameters
         ----------
         nan_as_null : bool, default False
-            Whether to tell the DataFrame to overwrite null values in the data
-            with ``NaN`` (or ``NaT``).
+            `nan_as_null` is DEPRECATED and has no effect. Please avoid using
+            it; it will be removed in a future release (after Aug 2024).
```

Comment:
The date is wrong, this will be removed in 3.0 which will happen before that.

Also: This needs a FutureWarning if a value is passed by the user and we need tests for that

Reply:
The intent here was to keep it around for a year in all dataframe libraries, to avoid unnecessary breakage. This works best if we remove usage of the keyword from all callers of `__dataframe__` now, and then from the signature a year later. 

Is there a reason to not do this? Removing it sooner seems to not have any upsides, only more disruption. I can remove the word "deprecated" from the PR if it bothers you?

> This needs a FutureWarning

This is not a good idea. The end user can do nothing about this. This is a between-dataframes communication mechanism, so emitting user-facing warnings has no purpose.

It's also not a behavior change but a removal of the keyword without any observable changes to how `__dataframe__` behavior, so if a warning were needed it should be a deprecation warning.

Reply:
We don't do API breaking changes in minor versions, so we can remove it in 3.0 or 4.0 (I don't have a preference), but not in a minor version in between (the date and lack of a warning bothers me, deprecated itself is fine).

You can start with a ``DeprecationWarning``, that is fine, but it needs a ``FutureWarning`` at some point before the keyword is actually removed. 

It is part of our API docs (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.__dataframe__.html), so there is no guarantee that it is only used by library developers and we have to consider this a public method.

Reply:
Hey all

I've looked at this a bit more - it's true that technically it may be meant to only be used between dataframe libraries in `from_dataframe`, but:
- as Patrick noted, it's part of the pandas API docs (and was added there as part of the PR which introduced the interchange protocol)
- there's a blog post which uses `DataFrame.__dataframe__` directy: https://www.thomasjpfan.com/2023/05/accessing-data-from-pythons-dataframe-interchange-protocol/

So, I'd suggest treating this as any other public-facing pandas method, and going through the usual deprecation process

In any case, I've had a look at https://github.com/search?q=__dataframe__%28&type=code&p=1, and it doesn't look like there's many users of `__dataframe__` anyway, and nobody seems to be passing anything non-default values to `nan_as_null` anyway. If we want to remove it, I'd suggest doing it earlier rather than later, before anyone gets the idea to try using it

> This is not a good idea.

generic communication comment - could we not make such statements about maintainers' reviews please? thanks 🙏

Reply:
Apologies for the too brief and too certain statement, I should have written _I don't think it is a good idea to emit ...".

I will note that the whole premise of the upstream change and the proposal by @jorisvandenbossche at https://github.com/data-apis/dataframe-api/issues/125 was that it can be removed because no one is using it - and hence that that is quick/safe/silent. A warning now would be really disruptive - so if you insist that the change cannot be made without one, it'd be better to close this PR and revisit it once all other dataframe libraries have stopped passing in this keyword in say 6-9 months.

Reply:
Thanks

Yeah I think that might be best to be honest - remove it from implementations and the official spec, and then do a proper deprecation cycle in pandas. There's a general desire in pandas to be stricter about the process around breaking changes now that there's a yearly release cycle, and I don't think that making an exception for `__dataframe__` would be seen too well

> it'd be better to close this PR

I don't think it needs closing - I think we can already note that the argument is deprecated and will be removed, and then, once the above has been completed, we follow the pandas deprecation process. So as this lives in pandas, then for better or for worse, that is what we're bound by

Reply:
@phofl gentle ping - OK to get this docs change in now and then do the usual deprecation once other implementations have been updated?

Reply:
Looking at the actual diff of the PR now: I think this is indeed fine to merge as is (maybe just remove the explicit date when it will be removed, for now). All it does is update the documentation to be clearer that this keyword doesn't do anything and is deprecated, and clean up the internals to already remove it internally (which has no user visible impact).